sys/linux/devio.txt

```diff
@@ -0,0 +1,129 @@
+
+mmap$devio(addr vma, len len[addr], prot flags[mmap_prot], flags flags[mmap_flags], fd fd_devio, offset fileoff)
+_ = __NR_mmap2
+
+ioctl$USBDEVFS_REAPURB(fd fd_devio, cmd const[USBDEVFS_REAPURB], arg ptr[out, int64])
+ioctl$USBDEVFS_REAPURBNDELAY(fd fd_devio, cmd const[USBDEVFS_REAPURBNDELAY], arg ptr[out, int64])
+ioctl$USBDEVFS_CONTROL(fd fd_devio, cmd const[USBDEVFS_CONTROL], arg ptr[in, usbdevfs_ctrltranfer])
+ioctl$USBDEVFS_BULK(fd fd_devio, cmd const[USBDEVFS_BULK], arg ptr[in, usbdevfs_bulktransfer])
+ioctl$USBDEVFS_RESETEP(fd fd_devio, cmd const[USBDEVFS_RESETEP], arg ptr[in, int32])
+ioctl$USBDEVFS_RESET(fd fd_devio, cmd const[USBDEVFS_RESET])
+ioctl$USBDEVFS_CLEAR_HALT(fd fd_devio, cmd const[USBDEVFS_CLEAR_HALT], arg ptr[in, int32])
+ioctl$USBDEVFS_GETDRIVER(fd fd_devio, cmd const[USBDEVFS_GETDRIVER], arg ptr[inout, usbdevfs_getdriver])
+ioctl$USBDEVFS_CONNECTINFO(fd fd_devio, cmd const[USBDEVFS_CONNECTINFO], arg ptr[out, usbdevfs_connectinfo])
+ioctl$USBDEVFS_SETINTERFACE(fd fd_devio, cmd const[USBDEVFS_SETINTERFACE], arg ptr[in, usbdevfs_setinterface])
+ioctl$USBDEVFS_SETCONFIGURATION(fd fd_devio, cmd const[USBDEVFS_SETCONFIGURATION], arg ptr[in, int32])
+ioctl$USBDEVFS_SUBMITURB(fd fd_devio, cmd const[USBDEVFS_SUBMITURB], arg ptr[in, usbdevfs_urb])
+ioctl$USBDEVFS_DISCARDURB(fd fd_devio, cmd const[USBDEVFS_DISCARDURB], arg ptr[in, int64])
+ioctl$USBDEVFS_DISCSIGNAL(fd fd_devio, cmd const[USBDEVFS_DISCSIGNAL], arg ptr[in, usbdevfs_disconnectsignal])
+ioctl$USBDEVFS_CLAIMINTERFACE(fd fd_devio, cmd const[USBDEVFS_CLAIMINTERFACE], arg ptr[in, int32])
+ioctl$USBDEVFS_RELEASEINTERFACE(fd fd_devio, cmd const[USBDEVFS_RELEASEINTERFACE], arg ptr[in, int32])
+ioctl$USBDEVFS_IOCTL(fd fd_devio, cmd const[USBDEVFS_IOCTL], arg ptr[in, usbdevfs_ioctl])
+ioctl$USBDEVFS_CLAIM_PORT(fd fd_devio, cmd const[USBDEVFS_CLAIM_PORT], arg ptr[in, int32])
+ioctl$USBDEVFS_RELEASE_PORT(fd fd_devio, cmd const[USBDEVFS_RELEASE_PORT], arg ptr[in, int32])
+ioctl$USBDEVFS_GET_CAPABILITIES(fd fd_devio, cmd const[USBDEVFS_GET_CAPABILITIES], arg ptr[out, int32])
+ioctl$USBDEVFS_DISCONNECT_CLAIM(fd fd_devio, cmd const[USBDEVFS_DISCONNECT_CLAIM], arg ptr[in, usbdevfs_disconnect_claim])
+ioctl$USBDEVFS_FREE_STREAMS(fd fd_devio, cmd const[USBDEVFS_FREE_STREAMS], arg ptr[in, usbdevfs_streams])
+ioctl$USBDEVFS_DROP_PRIVILEGES(fd fd_devio, cmd const[USBDEVFS_DROP_PRIVILEGES], arg ptr[in, int32])
+ioctl$USBDEVFS_GET_SPEED(fd fd_devio, cmd const[USBDEVFS_GET_SPEED])
+
+usbdevfs_ctrltranfer {
+	bRequestType	int8
+	bRequest	int8
+	wValue		int16
+	wIndex		int16
+	wLength		int16
+	timeout		int32
+	data		ptr[inout, array[int8]]
+}
+
+usbdevfs_bulktransfer {
+	ep	int32
+	len	len[data, int32]
+	timeout	int32
+	data	ptr[inout, array[int8]]
+}
+
+usbdevfs_getdriver {
+	interface	int32
+	driver		ptr[out, array[int8, MAXDRIVERNAME]]
+}
+
+usbdevfs_connectinfo {
+	devnum	int32
+	slow	int8
+}
+
+usbdevfs_setinterface {
+	interface	int32
+	altsetting	int32
+}
+
+usbdevfs_urb {
```

Comment:
I think we need to make this a union of 4 `usbdevfs_ubp_X` structs, each defined separately for each of the `urb_types` with more precise descriptions (e.g. for `USBDEVFS_URB_TYPE_CONTROL` the `buffer` field must point to a `usb_ctrlrequest` struct, etc.)